### ISSUE TITLE
Compile time formating checking(FMT_STRING)?

### ISSUE THREAD

**Issue by nanoric** [gabime/spdlog#1369]
fmt library support compile-time format error checking using FMT_STRING.
can spdlog enable this?

**Comment by gabime** [comment]
Maybe in one of the future versions.

**Comment by rabbull** [comment]
Seems current version of fmt embedded does not support compile-time format checking, so it's necessary to upgrade fmt to the latest version and test for compatibility first. If needed I'd like to prepaire a pr for it.

**Comment by gabime** [comment]
> Seems current version of fmt embedded does not support compile-time format checking, so it's necessary to upgrade fmt to the latest version

No sure I understand. spdlog already  uses the latest version of fmt - 6.1.2

**Comment by rabbull** [comment]
> > Seems current version of fmt embedded does not support compile-time format checking, so it's necessary to upgrade fmt to the latest version
> 
> No sure I understand. spdlog already uses the latest version of fmt - 6.1.2

Hmm that's awkward. Sry for the mistake. So is there anyone working on this feature? May I take this on?

**Comment by gabime** [comment]
Sure. help is welcome.

A possible  approach would to add a new macro in common.h that defaults to nothing or set to FMT_STRING by the user if desired. Then the fmt formatting code in logger.h would wrap the format string with this new macro.

**Comment by rabbull** [comment]
Bad news. I noticed all strings are finally converted into `string_view_t` before formatting, but the `FMT_STRING` macro does not support `string_view_t` (aka `fmt::basic_string_view<char>`) variable as its param.

Example:
```C++
#include <iostream>
#include <spdlog/spdlog.h>
int main() {
    using string_view_t = fmt::basic_string_view<char>;
    std::cout << fmt::format(FMT_STRING(string_view_t{"{}"}), 150) << std::endl;
}
```

While being compiled MinGW-w64 gives (haven't tried on the other compilers)
```
...
error: could not convert '{fmt::v6::basic_string_view<char>(((const char*)"{}")), ((sizeof (fmt::v6::basic_string_view<char>(((const char*)"{}"))) / sizeof (main(int, char**)::<lambda()>::str::char_type)) - 1)}' from '<brace-enclosed initializer list>' to 'fmt::v6::basic_string_view<fmt::v6::basic_string_view<char> >'
...
```

Inserting FMT_STRING into logger layer would trigger this error. So it'd be a huge patch to add this feature, either by changing interfaces (e.g. from the `string_view_t` way to the perfect forwarding way) or by implementing another `FMT_STRING`. 

BTW, an ok workaround is to emphasize `FMT_STRING` is user-visible in document, which could be invoked like:
```C++
spdlog::info(FMT_STRING("{:d} is an invalid tag"));
```

**Comment by gabime** [comment]
> BTW, an ok workaround is to emphasize FMT_STRING is user-visible in document, which could be invoked like..

Sounds good enough for me. Maybe even better than the original idea..

**Comment by gabime** [comment]
Added readme

**Comment by SX91** [comment]
@gabime hello! What was the point adding the line above to README if spdlog actually doesn't support fmtlib's compile-time checks?

```
#include "spdlog/spdlog.h"

int main()
{
//    fmt::format(FMT_STRING("{2}"), 123);
    spdlog::info(FMT_STRING("{2}"), 123);
    return 0;
}
```

The code above compiles perfectly fine with `fmt::format` call commented out.

**Comment by gabime** [comment]
You are right. It doesn't work. @rabbull ?

**Comment by SX91** [comment]
@gabime the type passed as format string to `fmt::*` functions should be derived from `fmt::compile_string` to trigger compile-time format string checks (look at `is_compile_string` uses). `spdlog` explicitly uses `string_view_t` type as format string thus disabling those checks by-design. So even if you use FMT_STRING macro, the result is implicitly converted to `string_view_t`. Oh, and `FMT_STRING` macro's `struct str` expects format string to be a static char string (`{s, sizeof(s) / sizeof(char_type) - 1}`).

**Comment by gabime** [comment]
Thanks. Removing this feature from the readme. 
Clearly more work need to be done for this to work

**Comment by gabime** [comment]
Its possible, but seems big change in the API to worth it.

**Comment by hak8or** [comment]
Can this be re-opened since, as @SX91 showed, this still does not actually work? Unless of course you feel that this should not be implemented due to the API change.

Somewhat a shame since one of the reasons I really liked fmt is due to the compile time checks. For example, I started sprinkling around things like ```spdlig::error("Oh no! foo()={:d}", rc);"```). But then I got burned later with a few of ```write(2, "[*** LOG ERROR #0001 ***] [2020-"..., 81[*** LOG ERROR #0001 ***] [2020-03-10 18:31:26] [] {argument index out of range}``` sprinkled around, hence googling and coming here. Though of course, this is your project so I respect whichever decision you make.

**Comment by gabime** [comment]
@hak8or its a shame I agree, I will take another look and check if there is a way to achieve this without changing the spdlog API so much.

@vitaut, your idea/recommendation for this would be greatly appreciated.

**Comment by vitaut** [comment]
The API for compile-time validation of format strings is currently internal (`internal::make_args_checked`) but I can make it public. The way to use it is to make format string type a template parameter (possibly constrained) and call `make_args_checked` instead of `make_format_args` in the top-level variadic function, for example:

```c++
template <typename S, typename... Args, typename Char = char_t<S>>
inline std::basic_string<Char> format(const S& format_str, Args&&... args) {
  return internal::vformat(
      to_string_view(format_str),
      internal::make_args_checked<Args...>(format_str, args...));
}
```

**Comment by gabime** [comment]
Currently spdlog does this:

```c++
    template<typename... Args>
    void log(source_loc loc, level::level_enum lvl, string_view_t fmt, const Args &... args)
    {
           ...
            memory_buf_t buf;
            fmt::format_to(buf, fmt, args...);
            ...
    }
```

I guess the idea is to replace it with something like the below?

```c++
    fmt::format_to(buf, fmt, make_args_checked<Args...>(fmt, args...)))
```

Would it work with `format_to`?

**Comment by vitaut** [comment]
You'll have to make the format string type a template parameter in `log`:

```c++
    template <typename S, typename... Args>
    void log(source_loc loc, level::level_enum lvl, const S& fmt, const Args &... args)
    {
           ...
            memory_buf_t buf;
            fmt::format_to(buf, fmt, args...);
            ...
    }
```

but otherwise it will work too (and doesn't need the internal API): https://godbolt.org/z/4YwqHA

**Comment by gabime** [comment]
> You'll have to make the format string type a template parameter
Right, however this is what I trying to avoid - changing the api..

**Comment by vitaut** [comment]
You can't do this without changing the API. This is a compatible change both source- and ABI-wise (at least in theory).

**Comment by vitaut** [comment]
I.e. all the code that works now, will continue to work after templatizing.

**Comment by vitaut** [comment]
Maybe except pathological cases such as taking an address of an instantiation of `log`.

**Comment by gabime** [comment]
Could a possible approach to avoid the api change is to i:
```c++
    fmt::format_to(buf, fmt, make_args_checked<Args...>(fmt, args...)))
```

instead?

**Comment by vitaut** [comment]
No, we have to smuggle a compile-time string as a type because C++ is a mess.

**Comment by skelly-energid** [comment]
@gabime There was a suggestion from @vitaut which is source compatible. This seems to be an improvement that wasn't made a few years ago but can be made now (https://github.com/gabime/spdlog/issues/338). 

Is there any reason not to make the change?

**Comment by gabime** [comment]
@skelly-energid . Perfect timing this was implemented today by @Tridacnid #1574

**Comment by gabime** [comment]
No. No need for this to achieve FMT_STRING
